### PR TITLE
lightning: fix risk of OOM

### LINKED ISSUE #40400: lightning: risk of OOM
# reproduce
1. create a csv file with unclosed quote till end:
```
1,2,3,"xxxxx...
```
2. insert 10GB random bytes into the data file
3. import

# result
9a0e5a85-6d7d-4e2f-a8d3-ed4b08fb6c8e

If user wants to import a big file with unclosed quote, lightning might possibly be OOM.

### PR DESCRIPTION
**Pull request by buchuitoudegou**
### What problem does this PR solve?

The risk of OOM when importing large data file
Issue Number: close #40400 

Problem Summary:

### What is changed and how it works?
- check if the buf has exceeded the max length of transaction
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/mydump/csv_parser.go [hunk 1]
```diff
@@ -336,6 +337,9 @@ func (parser *CSVParser) readUntil(chars *byteSet) ([]byte, byte, error) {
 	var buf []byte
 	for {
 		buf = append(buf, parser.buf...)
+		if len(buf) > tidbconfig.MaxTxnEntrySizeLimit {
```
Comment (gozssky) [inline br/pkg/lightning/mydump/csv_parser.go:340]: Prefer to do this check in `readRecord`.

Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/csv_parser.go:340]: But `readRecord` will invoke `readUntil`. If we check it in `readRecord` and there is no targeted byte in the following content, then too much memory will still be allocated for `buf`.

```
func (parser *CSVParser) readRecord(dst []string) ([]string, error) {
	parser.recordBuffer = parser.recordBuffer[:0]
	parser.fieldIndexes = parser.fieldIndexes[:0]

	isEmptyLine := true
	whitespaceLine := true
	prevToken := csvTokenNewLine
	var firstToken csvToken

outside:
	for {
		content, firstByte, err := parser.readUntil(&parser.unquoteByteSet)

```

Comment (gozssky) [reply br/pkg/lightning/mydump/csv_parser.go:340]: Oh, I missed it.

**Review by gozssky — APPROVED** [review]

**Comment by buchuitoudegou** [comment]
/cc @lance6716 

wait for BR's CI fixed.

File: br/pkg/lightning/mydump/csv_parser_test.go [hunk 2]
```diff
@@ -680,6 +682,28 @@ func TestConsecutiveFields(t *testing.T) {
...
+	cfg := config.MydumperRuntime{
+		CSV: config.CSVConfig{
+			Separator: ",",
+			Delimiter: `"`,
+		},
+	}
+	var testCase bytes.Buffer
+	testCase.WriteString("a,b,c,d")
+	// WARN: will take up 120MB memory here.
+	for i := 0; i < tidbconfig.MaxTxnEntrySizeLimit; i++ {
```
Comment (lance6716) [inline br/pkg/lightning/mydump/csv_parser_test.go:695]: maybe create a variable in `init` so we can easily test it. Not sure if CI can tolerate this.

**Review by lance6716 — APPROVED** [review]

File: br/pkg/lightning/mydump/csv_parser_test.go (hunk 2 above)
Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/csv_parser_test.go]: cd6e96e

Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/csv_parser_test.go]: dc59781

**Pull request merged**